src/ALE/PLM_functions.F90

```diff
@@ -5,21 +5,177 @@ module PLM_functions
+public PLM_boundary_extrapolation
+public PLM_extrapolate_slope
+public PLM_monotonized_slope
+public PLM_reconstruction
+public PLM_slope_wa
+public PLM_slope_cw
 
 real, parameter :: hNeglect_dflt = 1.E-30 !< Default negligible cell thickness
 
 contains
 
+!> Returns a limited PLM slope following White and Adcroft, 2008. [units of u]
+!! Note that this is not the same as the Colella and Woodward method.
+real elemental pure function PLM_slope_wa(h_l, h_c, h_r, h_neglect, u_l, u_c, u_r)
+  real, intent(in) :: h_l !< Thickness of left cell [units of grid thickness]
+  real, intent(in) :: h_c !< Thickness of center cell [units of grid thickness]
+  real, intent(in) :: h_r !< Thickness of right cell [units of grid thickness]
+  real, intent(in) :: h_neglect !< A negligible thickness [units of grid thickness]
+  real, intent(in) :: u_l !< Value of left cell [units of u]
+  real, intent(in) :: u_c !< Value of center cell [units of u]
+  real, intent(in) :: u_r !< Value of right cell [units of u]
+  ! Local variables
+  real :: sigma_l, sigma_c, sigma_r, u_min, u_max
+
+  ! Side differences
+  sigma_r = u_r - u_c
+  sigma_l = u_c - u_l
+
+  ! Quasi-second order difference
+  sigma_c = 2.0 * ( u_r - u_l ) * ( h_c / ( h_l + 2.0*h_c + h_r + h_neglect) )
+
+  ! Limit slope so that reconstructions are bounded by neighbors
+  u_min = min( u_l, u_c, u_r )
+  u_max = max( u_l, u_c, u_r )
+  if ( (sigma_l * sigma_r) > 0.0 ) then
+    ! This limits the slope so that the edge values are bounded by the
+    ! two cell averages spanning the edge.
+    PLM_slope_wa = sign( min( abs(sigma_c), 2.*min( u_c - u_min, u_max - u_c ) ), sigma_c )
+  else
+    ! Extrema in the mean values require a PCM reconstruction avoid generating
+    ! larger extreme values.
+    PLM_slope_wa = 0.0
+  endif
+
+  ! This block tests to see if roundoff causes edge values to be out of bounds
+  if (u_c - 0.5*abs(PLM_slope_wa) < u_min .or.  u_c + 0.5*abs(PLM_slope_wa) > u_max) then
+    PLM_slope_wa = PLM_slope_wa * ( 1. - epsilon(PLM_slope_wa) )
+  endif
+
+  ! An attempt to avoid inconsistency when the values become unrepresentable.
+  if (abs(PLM_slope_wa) < 1.E-140) PLM_slope_wa = 0.
+
+end function PLM_slope_wa
+
+!> Returns a limited PLM slope following Colella and Woodward 1984.
+real elemental pure function PLM_slope_cw(h_l, h_c, h_r, h_neglect, u_l, u_c, u_r)
+  real, intent(in) :: h_l !< Thickness of left cell [units of grid thickness]
+  real, intent(in) :: h_c !< Thickness of center cell [units of grid thickness]
+  real, intent(in) :: h_r !< Thickness of right cell [units of grid thickness]
+  real, intent(in) :: h_neglect !< A negligible thickness [units of grid thickness]
+  real, intent(in) :: u_l !< Value of left cell [units of u]
+  real, intent(in) :: u_c !< Value of center cell [units of u]
+  real, intent(in) :: u_r !< Value of right cell [units of u]
+  ! Local variables
+  real :: sigma_l, sigma_c, sigma_r, u_min, u_max, h_cn
```

Comment:
These local variables should have comments indicating their units, and variables with different units (i.e. h_cn) should be on a separate line if the variable units are being documented collectively.

Reply:
Addressed in 553166ae8e8a871a0d1e39a091cbe707ef8a3159